### PR TITLE
[ENH] Benchmarking using AptaNet and AptaTrans

### PR DESCRIPTION
**Pull request by satvshr**
solves #109

### CODE REVIEW

**Comment by satvshr** [comment]
@fkiraly @NennoMP for the AptaTrans dataset when I run a Benchmarking experiment on AptaNet it only predicts 0's (negatives), I know the dataset is imbalanced, but shouldn't it still be performing better than this?

When I run this script:
```
# --- imports
import pandas as pd
from sklearn.metrics import accuracy_score, f1_score
from sklearn.utils import resample

# AptaNetPipeline import (cover common locations)
from pyaptamer.aptanet import AptaNetPipeline
from pyaptamer.benchmarking import Benchmarking
from pyaptamer.benchmarking.preprocessors.aptanet import AptaNetPreprocessor
from pyaptamer.datasets import load_csv_dataset

# --- load raw dataset (aptamer/protein/y)
df = load_csv_dataset("train_li2014")

# --- preprocessor
pre = AptaNetPreprocessor()

# --- AptaNet model/pipeline (use defaults; tweak if you like)
clf = AptaNetPipeline()

# --- prepare labels
df = df.rename(columns={"label": "y"})
df["y"] = df["y"].map({"negative": 0, "positive": 1})

# =======================================================
# 1. Run benchmark on original (imbalanced) dataset
# =======================================================
bench = Benchmarking(
    estimators=[clf],
    evaluators=[accuracy_score, f1_score],
    task="classification",
    preprocessor=pre,
    datasets=df,  # can also be [df1, df2] or {"my_ds": df}
    test_size=0.25,
    stratify=True,
    random_state=1337,
)

results = bench.run()

# =======================================================
# 2. Perform class balancing (upsampling minority)
# =======================================================
df_majority = df[df.y == 0]
df_minority = df[df.y == 1]

df_minority_upsampled = resample(
    df_minority,
    replace=True,                 # sample with replacement
    n_samples=len(df_majority),   # match majority size
    random_state=1337
)

df_balanced = pd.concat([df_majority, df_minority_upsampled])
df_balanced = df_balanced.sample(frac=1, random_state=1337).reset_index(drop=True)

# =======================================================
# 3. Run benchmark again on balanced dataset
# =======================================================
bench_bal = Benchmarking(
    estimators=[clf],
    evaluators=[accuracy_score, f1_score],
    task="classification",
    preprocessor=pre,
    datasets=df_balanced,
    test_size=0.25,
    stratify=True,
    random_state=1337,
)

results_bal = bench_bal.run()
```
These are the results I get:
```
=== AptaNetPipeline ===
                                 dataset_0__test
train_set
dataset_0__train accuracy_score             0.75
                 f1_score                   0.00


=== AptaNetPipeline ===
                                 dataset_0__test
train_set
dataset_0__train accuracy_score         0.500000
                 f1_score               0.666667
```

**Comment by NennoMP** [comment]
> @fkiraly @NennoMP for the AptaTrans dataset when I run a Benchmarking experiment on AptaNet it only predicts 0's (negatives), I know the dataset is imbalanced, but shouldn't it still be performing better than this?

This is indeed very strange. While the dataset is a bit skewed I wouldn't expect such a behaviour. 

Could you check the predictions you obtain with an AptaNet which is untrained? In other words, you initialize an AptaNet with random weights and predict the labels. If everything is working correclty you should see both `1`'s and `0`'s. The model has not been trained yet and thus has no bias w.r.t. the majority/minority class. If you still observe all `0`'s predicted labels, there has to be a bug somewhere.

Let's do these checks for now.

> # =======================================================
> # 2. Perform class balancing (upsampling minority)
> # =======================================================
> df_majority = df[df.y == 0]
> df_minority = df[df.y == 1]
> 
> df_minority_upsampled = resample(
>     df_minority,
>     replace=True,                 # sample with replacement
>     n_samples=len(df_majority),   # match majority size
>     random_state=1337
> )
> 
> df_balanced = pd.concat([df_majority, df_minority_upsampled])
> df_balanced = df_balanced.sample(frac=1, random_state=1337).reset_index(drop=True)

Is there a reason why we are doing upsampling of the minority class here? Was it discussed in a weekly/daily?

I would use the dataset as it is (skewed) and simply make sure the validation split (if we use one during model selection) is obtained with stratification to ensure the same distribution.

**Comment by satvshr** [comment]
> Is there a reason why we are doing upsampling of the minority class here? Was it discussed in a weekly/daily?

It is a test script I wrote just for testing if I am getting only 0's for some reason, this is not a part of the algorithm.

**Comment by satvshr** [comment]
@NennoMP any idea why skorch maybe doing this?
Test:
```
import torch
import torch.nn as nn
import numpy as np
from skorch import NeuralNetBinaryClassifier
from pyaptamer.aptanet._aptanet_nn import AptaNetMLP  # adjust import if needed


def main():
    print("\n=== Forward pass without training (Torch vs Skorch) ===")

    # Shared dataset
    n_samples, n_features = 1000, 50
    X_np = np.random.randint(0, 10, size=(n_samples, n_features)).astype(np.float32)
    X_torch = torch.from_numpy(X_np)

    # --- Torch version ---
    model = AptaNetMLP()

    with torch.no_grad():
        logits = model(X_torch)
        probs = torch.sigmoid(logits)
        preds = (probs > 0.5).long()

    print("\n[Torch]")
    print("Pred labels distribution:", torch.bincount(preds.squeeze()).tolist())
    print("First 10 probabilities:\n", probs[:10].squeeze())

    # --- Skorch version ---
    net = NeuralNetBinaryClassifier(
        module=AptaNetMLP,
        module__input_dim=None,   # lazy
        module__hidden_dim=64,
        module__n_hidden=3,
        module__dropout=0.3,
        module__output_dim=1,
        module__use_lazy=True,
        criterion=nn.BCEWithLogitsLoss,
        max_epochs=0,             # no training
        device="cuda" if torch.cuda.is_available() else "cpu",
        verbose=0,
    )

    net.initialize()
    preds_skorch = net.predict(X_np)
    probs_skorch = net.predict_proba(X_np)

    print("\n[Skorch]")
    print("Pred labels distribution:", np.bincount(preds_skorch).tolist())
    print("First 10 probabilities:\n", probs_skorch[:10])


if __name__ == "__main__":
    main()
```

Results:
```
=== Forward pass without training (Torch vs Skorch) ===

[Torch]
Pred labels distribution: [450, 550]
First 10 probabilities:
 tensor([0.5068, 0.6765, 0.3494, 0.5657, 0.5299, 0.3275, 0.3092, 0.7680, 0.4583,
        0.3601])

[Skorch]
Pred labels distribution: [0, 1000]
First 10 probabilities:
 [[0.4526425  0.5473575 ]
 [0.46112305 0.53887695]
 [0.46430278 0.5356972 ]
 [0.4589256  0.5410744 ]
 [0.45875674 0.54124326]
 [0.46299422 0.5370058 ]
 [0.45827687 0.54172313]
 [0.46409988 0.5359001 ]
 [0.45826232 0.5417377 ]
 [0.46316904 0.53683096]]
```

**Comment by NennoMP** [comment]
> === Dataset Distribution ===
> Total samples: 1000
> Label distribution: [527, 473]
> 
> === Torch version ===
> [Before training]
> Pred labels distribution: [518, 482]
> First 10 probabilities:
>  tensor([0.5343, 0.4396, 0.6549, 0.3951, 0.5012, 0.4149, 0.5440, 0.2810, 0.6070,
>         0.4040])
> [After training]
> Pred labels distribution: [5, 995]
> First 10 probabilities:
>  tensor([0.5303, 0.5734, 0.5597, 0.5510, 0.5734, 0.5280, 0.5368, 0.5469, 0.5409,
>         0.5373])
> Accuracy: 0.478

If I read this correctly,  `torch` seems to work correctly when the model is not trained yet. We have approximately random guess `[518, 482]`, which is what should happen.

After training we are basically predicting the same class for nearly all samples. From this, I can only suppose the model is not training correctly.

> 
> === Skorch version ===
> [Before training]
> Pred labels distribution: [999, 1]
> First 10 probabilities:
>  [[0.5201893  0.47981068]
>  [0.5381153  0.46188468]
>  [0.51390857 0.48609143]
>  [0.5105337  0.48946628]
>  [0.52273107 0.4772689 ]
>  [0.5260486  0.4739514 ]
>  [0.5122141  0.48778588]
>  [0.5266731  0.4733269 ]
>  [0.51584256 0.48415747]
>  [0.50843424 0.49156576]]
> Re-initializing module because the following parameters were re-set: dropout, hidden_dim, input_dim, n_hidden, output_dim, use_lazy.
> Re-initializing criterion.
> Re-initializing optimizer.
>   epoch    train_loss    valid_acc    valid_loss     dur
> -------  ------------  -----------  ------------  ------
>       1        0.7199       0.4700        0.6972  0.0923
>       2        0.7111       0.4700        0.7079  0.0299
>       3        0.6953       0.4700        0.7038  0.0187
>       4        0.7110       0.5100        0.6921  0.0243
>       5        0.7007       0.4700        0.6995  0.0326
> [After training]
> Pred labels distribution: [0, 1000]
> First 10 probabilities:
>  [[0.43433756 0.56566244]
>  [0.42910838 0.5708916 ]
>  [0.43987745 0.56012255]
>  [0.45354545 0.54645455]
>  [0.4166276  0.5833724 ]
>  [0.4298364  0.5701636 ]
>  [0.43892968 0.5610703 ]
>  [0.4603036  0.5396964 ]
>  [0.45096928 0.5490307 ]
>  [0.4541608  0.5458392 ]]
> Accuracy: 0.473
> ```

`skorch` seems to have problems even when the model is not even trained ( `[999, 1]`)? Very strange, this shouldn't happen.
Also here, seems like the neural network isn't training at all, loss and accuracy are stale. There may be a bug in the training code or it could be a matter of hyperparameters.


I will try to look at the code or test it myself when I can.

**Review by fkiraly — CHANGES_REQUESTED** [review]
please make sure tests pass

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great start!

I think there are two issues that we should improve on here:

* that the dataset format for estimators and for the benchmark class is not the same, it should be the same convention for `X` and `y` throughout - now there is too much reformatting going on
* the "preprocessors" are in fact transformations that should be part of the algorithm, not done by the benchmark class. In the end, only the same type of estimator should get benchmarked.

Suggested actions:

* rework the class so data are directly passed as `X` and `y` (or collections thereof), or loaders
* rework the loaders that they return `X` and `y`-like objects
* for "raw" datasets, this should adopt the transformation input interface
* the "preprocessing" should be moved into pipelines with estimators, rather than handled by the benchmarking class

Further requests:
* change the benchmarking class to allow to pass `cv` objects from `sklearn` directly, do not hard code the validation strategy
* the `task` argument seems redundant, as it is implied by the data and the estimator, I would suggest to remove it
* the docstring is unclear as to what happens - there is no return or consequence described

File: pyaptamer/utils/tag_checks.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+"""Add all `_tag` checking functions here."""
+
+
+def task_check(estimator):
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply pyaptamer/utils/tag_checks.py:7]: can you explain what exactly this module is for, and where it is used?

[the remark this answers is not in the document]
Comment (satvshr) [reply pyaptamer/utils/tag_checks.py:7]: Given benchmarking is going to be used for both regression and classification tasks I added it to ensure no regression estimators s being used with a classification dataset and vice versa, should I remove it and assume the user knows what they are doing?

File: pyaptamer/benchmarking/_base.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        return pd.DataFrame(records, index=index, columns=["train", "test"])
+
+    def run(self):
+        """
+        Train each estimator and evaluate with hold-out or cross-validation.
+
+        Returns
+        -------
+        pd.DataFrame
+            Results table with rows = (estimator, metric),
```
[the remark this answers is not in the document]
Comment (satvshr) [reply pyaptamer/benchmarking/_base.py:176]: Added an example output

File: pyaptamer/datasets/_loaders/_csv_loader.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    sklearn.utils.Bunch or tuple of np.ndarray
+        If `return_X_y` is False, returns a Bunch with fields:
+            - data: ndarray of shape (n_samples, n_features)
+            - target: ndarray of shape (n_samples,)
+            - frame: pandas.DataFrame (the loaded DataFrame)
+            - feature_names: list[str] (column names used as features)
+            - target_name: str (the name of the target column)
+            - filename: str (resolved path to the CSV file)
+        If `return_X_y` is True, returns (X, y) where:
+            - X : ndarray of shape (n_samples, n_features) built by dropping the target
```
[the remark this answers is not in the document]
Comment (satvshr) [reply pyaptamer/datasets/_loaders/_csv_loader.py]: As discussed on [discord](https://discord.com/channels/1390583813414981755/1419586002879778816/1419587723563372625).

**Comment by fkiraly** [comment]
can you kindly summarize how requests were addressed and what the changes since last review are?

**Comment by satvshr** [comment]
The changes made were the ones requested, so:
1. The class now only take 3 arguments, X, y, and cv.
2. The csv loader returns dataframes instead of numpy arrays and bunches now.
3. Added an example to show what output the `run` method yields.
4. Added tests.

File: pyaptamer/benchmarking/_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply pyaptamer/benchmarking/_base.py:110]: please describe this better, say precisely which rows and columns there are and what the entries are

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

May I request to add a short notebook with a small benchmarking experiment and some reasonably chosen dataset?
I think that is important to to understand your design as well as for review.

**Comment by satvshr** [comment]
> In this one, it might allow to spot bugs which would otherwise need to be fixed in additional PR.

I can add bug fixing as part of the notebook PR as well.

**Comment by fkiraly** [comment]
could you kindly make sure you write a good PR description in the first post?

**Comment by satvshr** [comment]
> could you kindly make sure you write a good PR description in the first post?

Done.

**Review by fkiraly — CHANGES_REQUESTED** [review]
We are making changes to `AptaNetPipeline` - is this required for benchmarking? Would this not interact with other PRs, e.g., https://github.com/gc-os-ai/pyaptamer/pull/153?

I would recommend to make this a separate PR.

**Comment by NennoMP** [comment]
While investigating #142, I noticed that the benchmark class performs cross-validation on the given dataset (training data) to generate validation splits on which the model is then evaluated. However, the purpose of cross-validation is to maximize the use of training data during model selection (i.e., selection of the best hyperparameters), but we are not performing any model selection here.

Shouldn't the purpose of the benchmarking class be to train the model on training data using the best hyperparameters previously identified, and then perform a final evaluation on held-out test data (i.e., the `test_li2014.csv` file from AptaTrans)?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Requests have not been addressed. Repeating them here

* May I request to add a short notebook with a small benchmarking experiment and some reasonably chosen dataset? - first change request https://github.com/gc-os-ai/pyaptamer/pull/114#pullrequestreview-3281632768
    * as said there, this is necessary for me to understand your design
* I asked for the bugfix to be moved to a separate PR, and for a description of the bug you claim to fix. https://github.com/gc-os-ai/pyaptamer/pull/114#issuecomment-3383407987

**Comment by satvshr** [comment]
> you still have not explained what you think the bug is/was.

My bad, I realised I was not clear enough. `FunctionTransformer` takes a dictionary as input to `kw_args`, currently [it's](https://github.com/gc-os-ai/pyaptamer/blob/84c30e795894a7fcc60a605f56d9679fd8173fd9/pyaptamer/aptanet/_pipeline.py#L72)only taking a constant `k` as input.
The [tests](https://github.com/gc-os-ai/pyaptamer/blob/main/pyaptamer/aptanet/tests/test_aptanet.py) pass currently, as in the tests for `AptaNet`, `k` value is not is being passed to the `AptaNetPipeline`, which I also fix as part of [this](https://github.com/gc-os-ai/pyaptamer/blob/f9cece01931c1f05213ae870aa4bf825943d3f0f/pyaptamer/aptanet/tests/test_aptanet.py#L29) PR.

**Review by fkiraly — APPROVED** [review]
